### ISSUE TITLE
poetry develop failing on non-ASCII characters

### ISSUE THREAD

**Issue by xsduan** [python-poetry/poetry#221]
```toml
authors = [
    "Sébastien Eustace <sebastien@eustace.io>"
]
```

```bash
$ poetry develop -vvv

[AttributeError]
'NoneType' object has no attribute 'group'
```

```toml
authors = [
    "Sebastien Eustace <sebastien@eustace.io>"
]
```

```bash
Installing dependencies from lock file

Nothing to install or update

Installing poetry (0.11.0-alpha.3)
```

As far as I know, the `re` library doesn't have any ability to support unicode character classes but [`regex`](https://pypi.org/project/regex/) can handle them properly.

I don't know if this has been brought up before or this is a windows-only thing, considering this happened while `poetry develop`ing poetry itself. as far as I checked, nobody has made an issue about this before.

Windows 10, python 3.6.4, poetry 0.11.0a3. 

edit: #66 is similar.

In the meantime, catching errors:

```diff
    def _get_author(self):  # type: () -> dict
+       if self._authors:
+           m = AUTHOR_REGEX.match(self._authors[0])
+       else:
+           m = None

-       if not self._authors:
+       if not m:
+           # log.info('Could not find an author') or whatever
            return {"name": None, "email": None}

        m = AUTHOR_REGEX.match(self._authors[0])

        name = m.group("name")
        email = m.group("email")

        return {"name": name, "email": email}
```

**Comment by xsduan** [comment]
somehow the é is being encoded as iso latin-1 which is causing a unicode decode error.

```
    Complete output from command python setup.py egg_info:
    b"': 'S\xe9bast" # added print(data[24360:24370])
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "c:\users\droom\appdata\local\programs\python\python36\lib\codecs.py", line 331, in decode
        (result, consumed) = self._buffer_decode(data, self.errors, final)
    UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 24365: invalid continuation byte
```

**Comment by cauebs** [comment]
I remember having this problem, but I couldn't reproduce it now. On what version are you?

**Comment by xsduan** [comment]
poetry@0.11.2, ran `poetry develop -vvv`

**Comment by sdispater** [comment]
I only have this problem when the author name is retrieved from the Git config. If I set it manually by editing the `pyproject.toml` file directly or explicitely in the `init` command I don't have this issue.

**Comment by RaptDept** [comment]
I've done `poetry develop` on Poetry itself on this Windows box before without any issues, but I'm only getting this problem now. I'm not sure why I didn't run into this before.

https://github.com/sdispater/poetry/blob/a1b97707e8b193c3b3a7ee47394c155f9e1eb0c0/poetry/console/commands/develop.py#L31-L32

Since the `open()` call that creates `setup.py` doesn't explicitly specify an encoding, it falls back to [CP-1252](https://en.wikipedia.org/wiki/Windows-1252) encoding on my Windows system.

This conflicts with the `# -*- coding: utf-8 -*-` encoding declaration in format string used to create `setup.py`:

https://github.com/sdispater/poetry/blob/a1b97707e8b193c3b3a7ee47394c155f9e1eb0c0/poetry/masonry/builders/sdist.py#L25-L43

The simplest solution here is to specify `encoding="utf-8"` in the `open()` call.

For demonstration, running `poetry develop` on Poetry itself works in #368. If this looks okay, I'll write a test. There might be other places in the project that need to have explicit encodings, though -- I'm willing to take a look at that.

While looking into this issue, I also found that the same thing happens to the readme parsing:
https://github.com/sdispater/poetry/blob/a1b97707e8b193c3b3a7ee47394c155f9e1eb0c0/poetry/masonry/metadata.py#L48-L50

Poetry's README.md is UTF-8 on my machine but gets decoded as CP-1252 in `poetry develop`, turning all the `é`'s into `Ã©`, which in turn gets written out to the `description` field in `setup.py`. I'm not sure how this is supposed to be handled. Should it always assume that the readme file is UTF-8?

**Comment by xsduan** [comment]
I think utf8 is a reasonable assumption.

To be safe it could always retry after guessing the encoding with something like [chardet](https://pypi.org/project/chardet/) but I don't think that would be necessary, or at the very least just try CP-1252/ISO-8859 and then fail.

**Comment by vlcinsky** [comment]
As this issue is caused by `setup.py` being written in default encoding (and failing on systems, which have other than UTF-8 one), PR #1087 shall fix this issue (by being explicit about encoding when creating source-code like files)

**Comment by vlcinsky** [comment]
@xsduan Can you check, that the latest poetry 0.12.17 fixes the issue?

**Comment by xsduan** [comment]
```sh
# installation...
poetry master = $ cd d:\git\poetry
poetry master = $ poetry install
#...
poetry master = $ poetry run pip show poetry
Name: poetry
Version: 0.12.11
Summary: Python dependency management and packaging made easy.
Home-page: https://poetry.eustace.io/
Author: Sébastien Eustace
Author-email: sebastien@eustace.io
License: UNKNOWN
Location: d:\git\poetry
Requires: cachecontrol, cachy, cleo, html5lib, jsonschema, pkginfo, pyparsing, pyrsistent, requests-toolbelt, requests, shellingham, tomlkit
Required-by:
```

looks like it

**Comment by vlcinsky** [comment]
@xsduan it looks like ... your edit left your comment incomplete.

**Comment by laxas** [comment]
I can reproduce this error in poetry version 1.0.0 (Ubuntu 18.04)

```bash
>> poetry init --author "Alex Müller"
...
Package name [test]:  
Version [0.1.0]:  
Description []:  
[UnicodeDecodeError]
'ascii' codec can't decode byte 0xc3 in position 6: ordinal not in range(128)
```
It seams that "poetry init" can't handle the non ascii character in the author default in the dialogue. However it has no problem if the non ascii character is put after the prompt. 

The following works fine
```bash
>> poetry init --author Alex`
...
Package name [test]:  
Version [0.1.0]:  
Description []:  
Author [Alex, n to skip]:  Alex Müller
License []:  
```

**Comment by finswimmer** [comment]
Thanks @laxas ,

with your example I was able to reproduce it with python 2.7. With python3 it works.

The problem seems to be well known, e.g. : https://stackoverflow.com/questions/21129020/how-to-fix-unicodedecodeerror-ascii-codec-cant-decode-byte

A simple fix would be changing this:

https://github.com/python-poetry/poetry/blob/affe32d8d41c76b5cb908fca492d36b6cebb0f76/poetry/console/commands/init.py#L81

into

```
name = self.option("name")

if isinstance(name, str):
    name = name.decode().encode("UTF-8")
```

But I guess, this is such a general problem and should be fixed in another place.

fin swimmer

**Comment by miedzinski** [comment]
I have the same problem as @laxan. Can't init new project because of non-ascii character in my git user name.

[![asciicast](https://asciinema.org/a/SXKsE1lP8ctQ7oRkkpSlbgGqp.svg)](https://asciinema.org/a/SXKsE1lP8ctQ7oRkkpSlbgGqp)

**Comment by Abuelodelanada** [comment]
> I have the same problem as @laxan. Can't init new project because of non-ascii character in my git user name.
> 
> [![asciicast](https://camo.githubusercontent.com/161f2db388af257b0249b2dfe6bf72d2de4ee1e6/68747470733a2f2f61736369696e656d612e6f72672f612f53584b7345316c5038637451376f526b6b70536c62674771702e737667)](https://asciinema.org/a/SXKsE1lP8ctQ7oRkkpSlbgGqp)

The same here! My lastname has an **ó**


```bash
➜  rgh git:(develop) poetry  init -vvv

This command will guide you through creating your pyproject.toml config.

Package name [rgh]:
Version [0.1.0]:
Description []:
'ascii' codec can't encode character u'\xf3' in position 29: ordinal not in range(128)
'ascii' codec can't encode character u'\xf3' in position 29: ordinal not in range(128)
'ascii' codec can't encode character u'\xf3' in position 29: ordinal not in range(128)
...
```

**Comment by jmfederico** [comment]
Passing the author to the cli with non-ascii characters also triggers the error:

```bash
❯ poetry init --author="an accent í is non-ascii"

This command will guide you through creating your pyproject.toml config.

Package name [non-ascii-test]:
Version [0.1.0]:
Description []:

[UnicodeDecodeError]
'ascii' codec can't decode byte 0xc3 in position 10: ordinal not in range(128)
```

But if a non-ascii character is used when prompted to confirm the author, it does not fail:

```bash
❯ poetry init --author="ascii"

This command will guide you through creating your pyproject.toml config.

Package name [non-ascii-test]:
Version [0.1.0]:
Description []:
Author [ascii, n to skip]:  non-ascii í
License []:
...
```

**Comment by julienmalard** [comment]
I get the same from non-ascii package names, for example:

```
[tool.poetry]
name = "lassi"
version = "0.1.0"
description = ""
authors = ["ਜ਼ੂਲੀਏਂ ਮਲਾਰ (Julien Malard) <julien.malard@mail.mcgill.ca>"]
packages = [
    { include = "ਲੱਸੀ" }
]
```

**Comment by jacebrowning** [comment]
Poetry `1.0.10` now displays a slightly more clear error message:

```
$ poetry build
Building lassi (0.1.0)

[ValueError]
Invalid author string. Must be in the format: John Smith <john@example.com>
```

using this minimal `pyproject.toml`:

```
[tool.poetry]
name = "lassi"
version = "0.1.0"
description = ""
authors = ["ਜ਼ੂਜ਼ੂਜ਼ੂ ਜ਼ੂਜ਼ੂਜ਼ੂ <user@example.com>"]
```

**Comment by julienmalard** [comment]
@jacebrowning Thank you! If you could point me to the place where the AUTHOR_REGEX is defined (I can't find it!) I would be happy to contribute a pull request to help fix this. Perhaps a good approach would be to validate only what is in between <> tags, and allow the name to be anything? (Because some languages will use apostrophes, colons, combining makrs and other characters that re is likely to miss?)

**Comment by jacebrowning** [comment]
@julienmalard It looks like `AUTHOR_REGEX ` is now part of Poetry Core: https://github.com/python-poetry/poetry-core/search?q=AUTHOR_REGEX&unscoped_q=AUTHOR_REGEX

And imported here: https://github.com/python-poetry/poetry/blob/d2fd581c9a856a5c4e60a25acb95d06d2a963cf2/poetry/console/commands/init.py#L491-L498

**Comment by julienmalard** [comment]
@jacebrowning Thank you! I had not noticed that poetry.core was not part of this repository.

**Comment by stanislaw** [comment]
I have just encountered this problem when I tried installing and running Poetry from my raw Docker container with Ubuntu Bionic.

The quick fix for me was to do:

```
LC_ALL=C.UTF-8 poetry
```

A more permanent solution for my Docker container I have found [here](https://stackoverflow.com/a/31987791/598057):

```
sudo apt-get -y install language-pack-en
The following extra packages will be installed:
  language-pack-en-base
Generating locales...
  en_GB.UTF-8... /usr/sbin/locale-gen: done
Generation complete.
```

**Comment by vlcinsky** [comment]
@stanislaw what version of poetry do you have?

What you describe is a workaround.

`poetry` shall not depend on current settings for locales etc., it shall explicitly work with utf-8. If this is still not true, the fix (in poetry code) is to explicitly specify encoding utf-8 with all file open operations.

**Comment by stanislaw** [comment]
@vlcinsky sure I understand. I just needed to get something done really quickly.

```
root@95eea793181d:/app# poetry --version
Poetry version 1.1.4
```

**Comment by vlcinsky** [comment]
Thanks @stanislaw for detailed report.

My note was about robustness of `poetry`. Workarounds are very practical and often life savers.

I have met dependency of `setuptools` on current system locale which is definitely wrong (I think my fix is already in). Solution is searching through the project for all `open` calls and making sure, that if opening a stream in text mode, they do explicitly state the encoding "utf-8".

**Comment by julienmalard** [comment]
@jacebrowning Thanks for the pointed regarding `AUTHOR_REGEX`. After a bit of experimentation, I think that this has to do not with Poetry *per se* but rather with a bug in the `re` module (see https://github.com/lark-parser/lark/issues/590).

Replacing `re` with `regex` solves everything:
```
import re
AUTHOR_REGEX = re.compile(r"(?u)^(?P<name>[- .,\w\d'’\"()]+) <(?P<email>.+?)>$")
AUTHOR_REGEX.match("ம. ஆ. ஜூலீஎன் <julien.malard@mail.mcgill.ca>")
>>> None
# But...
import regex as re
AUTHOR_REGEX = re.compile(r"(?u)^(?P<name>[- .,\w\d'’\"()]+) <(?P<email>.+?)>$")
AUTHOR_REGEX.match("ம. ஆ. ஜூலீஎன் <julien.malard@mail.mcgill.ca>")
>>> <regex.Match object; span=(0, 44), match='ம. ஆ. ஜூலீஎன் <julien.malard@mail.mcgill.ca>'>
```

So my question would now be - should I submit a pull request with `import regex as re` to Poetry? Or would adding a dependency risk breaking things?
Thanks!
